File: java/socketd/src/main/java/org/noear/socketd/transport/core/entity/EntityDefault.java
package org.noear.socketd.transport.core.entity;

import org.noear.socketd.transport.core.Constants;
import org.noear.socketd.transport.core.Entity;
import org.noear.socketd.transport.core.EntityMetas;
import org.noear.socketd.utils.StrUtils;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 实体默认实现
 *
 * @author noear
 * @since 2.0
 */
public class EntityDefault implements Entity {

    private Map<String, String> metaMap;
    private String metaString = Constants.DEF_META_STRING;
    private boolean metaStringChanged = false;
    private ByteBuffer data = Constants.DEF_DATA;
    private int dataSize = 0;

    /**
     * At player name
     */
    public EntityDefault at(String name) {
        metaPut("@", name);
        return this;
    }

    /**
     * Range
     */
    public EntityDefault range(int start, int size) {
        metaPut(EntityMetas.META_RANGE_START, String.valueOf(start));
        metaPut(EntityMetas.META_RANGE_SIZE, String.valueOf(size));
        return this;
    }

    /**
     * 设置元信息字符串
     */
    public EntityDefault metaStringSet(String metaString) {
        this.metaMap = null;
        this.metaString = metaString;
        this.metaStringChanged = false;
        return this;
    }

    /**
     * 获取元信息字符串（queryString style）
     */
    @Override
    public String metaString() {
        if (metaStringChanged) {
            StringBuilder buf = new StringBuilder();

            List<String> metaKeys = new ArrayList<>(metaMap().keySet());

            for (String name : metaKeys) {
                String val = metaMap().get(name);
                buf.append(name).append("=").append(val).append("&");
            }

            if (buf.length() > 0) {
                buf.setLength(buf.length() - 1);
            }

            metaString = buf.toString();
            metaStringChanged = false;
        }

        return metaString;
    }

    /**
     * 放置元信息字典
     *
     * @param metaMap 元信息字典
     */
    public EntityDefault metaMapPut(Map<String, String> metaMap) {
        if (metaMap != null && metaMap.size() > 0) {
            this.metaMap().putAll(metaMap);
            this.metaStringChanged = true;
        }
        return this;
    }

    /**
     * 获取元信息字典
     */
    @Override
    public Map<String, String> metaMap() {
        if (metaMap == null) {
            metaMap = new ConcurrentHashMap<>();
            metaStringChanged = false;

            //此处要优化
            if (StrUtils.isNotEmpty(metaString)) {
                for (String kvStr : metaString.split("&")) {
                    int idx = kvStr.indexOf('=');
                    if (idx > 0) {
                        metaMap.put(kvStr.substring(0, idx), kvStr.substring(idx + 1));
                    }
                }
            }
        }

        return metaMap;
    }

    /**
     * 放置元信息
     *
     * @param name 名字
     * @param val  值
     */
    public EntityDefault metaPut(String name, String val) {
        if (val == null) {
            metaMap().remove(name);
        } else {
            metaMap().put(name, val);
        }

        this.metaStringChanged = true;
        return this;
    }

    /**
     * 删除元信息
     *
     * @param name 名字
     */
    public EntityDefault metaDel(String name) {
        metaMap().remove(name);
        this.metaStringChanged = true;
        return this;
    }

    /**
     * 获取元信息
     *
     * @param name 名字
     */
    @Override
    public String meta(String name) {
        return metaMap().get(name);
    }

    /**
     * 获取元信息或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    @Override
    public String metaOrDefault(String name, String def) {
        return metaMap().getOrDefault(name, def);
    }

    /**
     * 放置元信息
     *
     * @param name 名字
     * @param val  值
     */
    @Override
    public void putMeta(String name, String val) {
        metaPut(name, val);
    }

    /**
     * 删除元信息
     * */
    @Override
    public void delMeta(String name) {
        metaDel(name);
    }

    /**
     * 设置数据
     *
     * @param data 数据
     */
    public EntityDefault dataSet(byte[] data) {
        this.data = ByteBuffer.wrap(data);
        this.dataSize = data.length;
        return this;
    }

    /**
     * 设置数据
     *
     * @param data 数据
     */
    public EntityDefault dataSet(ByteBuffer data) {
        this.data = data;
        this.dataSize = data.limit();
        return this;
    }

    /**
     * 获取数据（若多次复用，需要reset）
     */
    @Override
    public ByteBuffer data() {
        return data;
    }

    /**
     * 获取数据并转成字符串
     */
    @Override
    public String dataAsString() {
        if (dataAsString == null) {
            dataAsString = new String(dataAsBytes(), StandardCharsets.UTF_8);
        }

        return dataAsString;
    }

    private String dataAsString;

    @Override
    public byte[] dataAsBytes() {
        if (data instanceof MappedByteBuffer) {
            byte[] tmp = new byte[dataSize];
            data.mark();
            data.get(tmp);
            data.reset();
            return tmp;
        } else {
            return data.array();
        }
    }

    /**
     * 获取数据长度
     */
    @Override
    public int dataSize() {
        return dataSize;
    }

    /**
     * 释放资源
     */
    @Override
    public void release() throws IOException {

    }

    @Override
    public String toString() {
        return "Entity{" +
                "meta='" + metaString() + '\'' +
                ", data=byte[" + dataSize + ']' + //避免内容太大，影响打印
                '}';
    }
}

File: python/socket_websocket/socketd/core/module/EntityDefault.py
from abc import ABC
import pickle

from .Entity import Entity


class EntityDefault(Entity, ABC):
    def __init__(self):
        self.meta_map = None
        self.meta_string = "_dEF__mET_a__sTRING"
        self.meta_stringChanged = False
        self.data: bytes = None
        self.data_size = 0

    def set_meta_string(self, meta_string):
        self.meta_map = None
        self.meta_string = meta_string
        self.meta_stringChanged = False
        return self

    def get_meta_string(self):
        if self.meta_stringChanged:
            buf = ""
            for name, val in self.get_meta_map().items():
                buf += f"{name}={val}&"
            if len(buf) > 0:
                buf = buf[:-1]
            self.meta_string = buf
            self.meta_stringChanged = False
        return self.meta_string

    def set_meta_map(self, meta_map):
        self.meta_map = meta_map
        self.meta_string = None
        self.meta_stringChanged = True
        return self

    def get_meta_map(self):
        if self.meta_map is None:
            self.meta_map = {}
            self.meta_stringChanged = False
            if self.meta_string:
                for kv_str in self.meta_string.split("&"):
                    kv = kv_str.split("=")
                    if len(kv) > 1:
                        self.meta_map[kv[0]] = kv[1]
                    else:
                        self.meta_map[kv[0]] = ""
        return self.meta_map

    def set_meta(self, name, val):
        self.put_meta(name, val)
        return self

    def put_meta(self, name, val):
        self.get_meta_map()[name] = val
        self.meta_stringChanged = True

    def get_meta(self, name):
        return self.get_meta_map().get(name)

    def get_metaOr_default(self, name, default_val):
        return self.get_meta_map().get(name, default_val)

    def set_data(self, data):
        if type(data) != bytes:
            self.data = pickle.loads(data)
        else:
            self.data = data
        self.data_size = len(data)
        return self

    def get_data(self):
        return self.data

    def get_data_as_string(self):
        return str(self.data, 'utf-8')  # _assuming data is of type bytes

    def get_data_size(self):
        return self.data_size

    def __str__(self):
        return f"Entity(meta='{self.get_meta_string()}', data=byte[{self.data_size}])"


